pkg/virt-handler/vm.go

```diff
@@ -2341,6 +2341,7 @@ func (d *VirtualMachineController) vmUpdateHelperMigrationSource(origVMI *v1.Vir
 				return err
 			}
 			d.recorder.Event(vmi, k8sv1.EventTypeNormal, v1.Migrating.String(), "VirtualMachineInstance is aborting migration.")
+			d.finalizeMigration(vmi)
```

Comment:
If I read this right, the `finalizeMigration` is to be executed when `AbortRequested`, therefore this is done when the virt-controller updates the flag.

First, I am not sure if we may pass though this code flow more than once.
Based on the [cancel function](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L322) at the virt-launcher, this portion of the code is only relevant to explicitly abort a running migration and not to do so on an already failed one. Your addition now changes this.

We have these scenario I know of:
- Setup of the migration failed (migration was not started yet on the domain itself).
- Migration failed, sourcing from libvirt itself.
- Migration failed, sourcing from the monitoring go-routine at the source node.
- Migration failed because the virt-controller decided so (or because a user asked to cancel).

All of these are handled here?
For example, what happens if the monitor marks the migration as failed, without an abort reason? Is that propagated to the virt-handler and then to the virt-controller which in turn sets this `AbortRequested` to `true`?

Reply:
@EdDev 
> If I read this right, the `finalizeMigration` is to be executed when `AbortRequested`, therefore this is done when the virt-controller updates the flag.
> 

True

> First, I am not sure if we may pass though this code flow more than once.

If you mean that due to an VMI update issue we will get there again then, yes this is what going to happen.

Upper the call stack, `defaultExecute` call's this function and eventually updates the VMI status as part of VMI sync.
In case of an error the request will re-queued and this code will run again.

> Based on the [cancel function](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L322) at the virt-launcher, this portion of the code is only relevant to explicitly abort a running migration and not to do so on an already failed one. Your addition now changes this.
> 
What does it changes? you think its not contradicts with cancel logic on virt-launcher?

> We have these scenario I know of:
> 
> * (1) Setup of the migration failed (migration was not started yet on the domain itself).
> * (2) Migration failed, sourcing from libvirt itself.
> * (3) Migration failed, sourcing from the monitoring go-routine at the source node.
> * (4) Migration failed because the virt-controller decided so (or because a user asked to cancel).
> 
> All of these are handled here?

(4) - handled here.
(1), (2) and (3) - are handled at line  [[1]](https://github.com/kubevirt/kubevirt/pull/5922/files/8d060997ef64c86ccb03c8480912d429e25b318c..f73036acf559f4cf05f9e1de15e44d1a0b6be667#diff-7a3f9e0461f554b07ac195375eb0f8db7d46579071795a65c459c721b1a8204aR2360).
Note about (3), in case of an error it will update the `migration.Status` that is handled on virt-handler code.

> For example, what happens if the monitor marks the migration as failed, without an abort reason? Is that propagated to the virt-handler and then to the virt-controller which in turn sets this `AbortRequested` to `true`?

In this case it means that the migration was failed and not canceled by the user,
it will propagate to virt-handler as failing migration, vmi.Status.MigrationState.abortRequestd and abortStatus are not in the mix in this case.

Form this line here is how it goes:
Its the virt-controller that set the migration object with AbortRequested = `true`
then virt-handler handles this abort request by sending cancel-migration and finalize-migration requests 
to virt-launcher 
virt-launcher cancelMigration logic runs, which sets the domainSpec with Migration.AbortStatus = 'Aborting'
and cancel Libvirt migration job.
* on the background virt-handler updates the vmi.Status.MigrationState according to the domain migration metadata 
migration-monitor finds that the migration job was canceled and set the domainSpec with
  Completed: true, Failed: true, Phase: Succeeded, Reason: Live migration aborted
* on the background virt-handler updates the vmi.Status.MigrationState ...

[1] https://github.com/kubevirt/kubevirt/pull/5922/files/8d060997ef64c86ccb03c8480912d429e25b318c..f73036acf559f4cf05f9e1de15e44d1a0b6be667#diff-7a3f9e0461f554b07ac195375eb0f8db7d46579071795a65c459c721b1a8204aR2360

Reply:
I edited my last comment with few refinements ^^